### PR TITLE
refactor: introduce `proto_reader.rs`

### LINKED ISSUE #1619: refactor: create `proto_reader`
Once https://github.com/tailcallhq/tailcall/issues/1618 is done, refactor `config::reader` and pull out the proto reading logic. The proto reader should take in a list of paths and return a `Vec<FileDescriptorSet>`.

### PR DESCRIPTION
**Pull request by ssddOnTop**
**Issue Reference(s):**  
Fixes #1619

## Summary by CodeRabbit

- **New Features**
	- Introduced a new way to resolve and parse Protocol Buffer (protobuf) files, including handling nested imports and well-known Google proto files.
- **Refactor**
	- Revamped the logic for handling protobuf links and grpc file descriptors, enhancing efficiency and clarity.
- **Tests**
	- Added testing for the resolution of proto files and their nested imports.

### CODE REVIEW

File: src/proto_reader.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    pub package: Option<String>,
+    pub descriptor_set: FileDescriptorSet,
+}
+
+impl ProtoReader {
+    pub fn init(runtime: TargetRuntime) -> Self {
+        Self { resource_reader: ResourceReader::init(runtime) }
+    }
+
+    pub async fn resolve_protos<T: AsRef<str>>(
```
Comment (tusharmath) [inline src/proto_reader.rs:25]: doesn't need to be public.

Comment (tusharmath) [reply src/proto_reader.rs:25]: Should only have `read` and `read_all` methods as public.

Comment (ssddOnTop) [reply src/proto_reader.rs:25]: this is the new function I introduced to resolve all the protos.. we can directly use this in config reader and generator reader without having to worry about parent and resolved descriptors

File: src/proto_reader.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        &self,
+        parent_proto: FileDescriptorProto,
+    ) -> anyhow::Result<Vec<FileDescriptorProto>> {
+        let mut descriptors: HashMap<String, FileDescriptorProto> = HashMap::new();
+        let mut queue = VecDeque::new();
+        queue.push_back(parent_proto.clone());
+
+        while let Some(file) = queue.pop_front() {
+            for import in file.dependency.iter() {
+                let proto = self.read_proto(import).await?;
```
[the remark this answers is not in the document]
Comment (ssddOnTop) [reply src/proto_reader.rs:58]: can we?

File: src/config/reader.rs [hunk 3]
```diff
@@ -57,6 +57,44 @@ impl ConfigReader {
...
+                format!(
+                    "Package name is not defined for proto file: {:?} with link id: {:?}",
+                    file_descriptor_set
+                        .descriptor_set
+                        .file
+                        .first()
+                        .unwrap()
+                        .name
+                        .as_ref()
+                        .unwrap(),
```
Comment (tusharmath) [inline src/config/reader.rs:82]: Unwrap will panic.

File: src/config/reader.rs [hunk 4]
```diff
@@ -57,6 +57,44 @@ impl ConfigReader {
...
         }
 
+        let protobuf_links = links
+            .iter()
+            .filter(|v| v.type_of == LinkType::Protobuf)
+            .collect::<Vec<_>>();
+        let protobuf_links_paths = protobuf_links
+            .iter()
+            .map(|v| Self::resolve_path(&v.src, parent_dir))
+            .collect::<Vec<String>>();
```
Comment (tusharmath) [inline src/config/reader.rs:67]: Don't convert to iter and then to vec and then back to iter. Perform all the operations while we have an iter.

File: src/config/reader.rs [hunk 5]
```diff
@@ -57,6 +57,35 @@ impl ConfigReader {
...
+        let file_descriptor_metadata = self.proto_reader.read_all(&protobuf_links_paths).await?;
+        for (i, file_descriptor_set) in file_descriptor_metadata.iter().enumerate() {
+            let id = Valid::from_option(
```
Comment (tusharmath) [inline src/config/reader.rs:70]: don't use Valid here.

**Review by tusharmath — APPROVED** [review]

**Pull request merged**